presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPageSink.java

```diff
@@ -296,27 +357,53 @@ private int[] getWriterIndexes(Page page)
         Page transformedPage = pagePartitioner.getTransformedPage();
         for (int position = 0; position < page.getPositionCount(); position++) {
             int writerIndex = writerIndexes[position];
-            if (writers.get(writerIndex) != null) {
+            WriteContext writer = writers.get(writerIndex);
+            if (writer != null) {
                 continue;
             }
 
             Optional<PartitionData> partitionData = getPartitionData(pagePartitioner.getColumns(), transformedPage, position);
-            WriteContext writer = createWriter(partitionData);
 
-            writers.set(writerIndex, writer);
+            String fileName = fileFormat.addExtension(randomUUID().toString());
+            Path outputPath = partitionData.map(partition -> new Path(locationProvider.newDataLocation(partitionSpec, partition, fileName)))
+                    .orElse(new Path(locationProvider.newDataLocation(fileName)));
+
+            try {
+                FileSystem fileSystem = hdfsEnvironment.getFileSystem(session.getUser(), outputPath, jobConf);
+                if (!sortOrder.isEmpty() && sortedWritingEnabled) {
+                    Path tempFilePrefix = new Path(tempDirectory, format("sorting-file-writer-%s-%s", session.getQueryId(), randomUUID()));
+                    WriteContext writerContext = createWriter(partitionData, outputPath);
+                    IcebergFileWriter sortedFileWriter = new IcebergSortingFileWriter(
+                            fileSystem,
+                            tempFilePrefix,
+                            writerContext.getWriter(),
+                            sortingFileWriterBufferSize,
+                            sortingFileWriterMaxOpenFiles,
+                            columnTypes,
+                            sortColumnIndexes,
+                            sortOrders,
+                            pageSorter,
+                            false,
+                            session,
+                            orcFileWriterFactory);
+                    writer = new WriteContext(sortedFileWriter, outputPath, partitionData);
+                }
+                else {
+                    writer = createWriter(partitionData, outputPath);
+                }
+                writers.set(writerIndex, writer);
+            }
+            catch (IOException e) {
+                throw new PrestoException(HIVE_WRITER_OPEN_ERROR, e);
```

Comment:
Is it appropriate to throw `HIVE_WRITER_OPEN_ERROR`? Maybe something like `SORT_WRITER_OPEN_ERROR`?